homeassistant/components/snapcast/media_player.py

```diff
@@ -61,51 +77,111 @@ async def async_setup_entry(
 
-    register_services()
+    # Fetch coordinator from global data
+    coordinator: SnapcastUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
 
+    # Create an ID for the Snapserver
     host = config_entry.data[CONF_HOST]
     port = config_entry.data[CONF_PORT]
-    hpid = f"{host}:{port}"
+    host_id = f"{host}:{port}"
 
-    groups: list[MediaPlayerEntity] = [
-        SnapcastGroupDevice(group, hpid, config_entry.entry_id)
-        for group in snapcast_server.groups
-    ]
-    clients: list[MediaPlayerEntity] = [
-        SnapcastClientDevice(client, hpid, config_entry.entry_id)
-        for client in snapcast_server.clients
-    ]
-    async_add_entities(clients + groups)
-    hass.data[DOMAIN][
-        config_entry.entry_id
-    ].hass_async_add_entities = async_add_entities
+    register_services()
 
+    _known_group_ids: set[str] = set()
+    _known_client_ids: set[str] = set()
 
-async def handle_async_join(entity, service_call):
-    """Handle the entity service join."""
-    if not isinstance(entity, SnapcastClientDevice):
-        raise TypeError("Entity is not a client. Can only join clients.")
-    await entity.async_join(service_call.data[ATTR_MASTER])
+    @callback
+    def _check_entities() -> None:
+        nonlocal _known_group_ids, _known_client_ids
 
+        def _update_known_ids(known_ids, ids) -> tuple[str, str]:
+            ids_to_add = ids - known_ids
+            ids_to_remove = known_ids - ids
 
-async def handle_async_unjoin(entity, service_call):
-    """Handle the entity service unjoin."""
-    if not isinstance(entity, SnapcastClientDevice):
-        raise TypeError("Entity is not a client. Can only unjoin clients.")
-    await entity.async_unjoin()
+            # Update known IDs
+            known_ids -= ids_to_remove
+            known_ids |= ids_to_add
 
+            return ids_to_add, ids_to_remove
 
-async def handle_set_latency(entity, service_call):
-    """Handle the entity service set_latency."""
-    if not isinstance(entity, SnapcastClientDevice):
-        raise TypeError("Latency can only be set for a Snapcast client.")
-    await entity.async_set_latency(service_call.data[ATTR_LATENCY])
+        new_entities: list[SnapcastBaseDevice] = []
 
+        group_ids = {g.identifier for g in coordinator.server.groups}
+        groups_to_add, groups_to_remove = _update_known_ids(_known_group_ids, group_ids)
 
-class SnapcastGroupDevice(MediaPlayerEntity):
-    """Representation of a Snapcast group device."""
+        new_entities.extend(
+            [
+                SnapcastGroupDevice(
+                    coordinator, coordinator.server.group(group_id), host_id
+                )
+                for group_id in groups_to_add
+            ]
+        )
+
+        client_ids = {c.identifier for c in coordinator.server.clients}
+        clients_to_add, clients_to_remove = _update_known_ids(
+            _known_client_ids, client_ids
+        )
```

Comment:
After this point we know if we have new or to remove devices. We should rather quick return here if we have no action to do (unless this code is only executed when a player/group connects/disconnects)

Reply:
The code gets executed whenever there is a server update. @luar123 had some concerns about that in another comment too.

Implemented the early exit after the sets are updated.